### ISSUE TITLE
issue with 'gettimeofday' for event loop time calculation

### ISSUE THREAD

**Issue by chenbd** [h2o/h2o#2321]
Currently **gettimeofday** is used for h2o event loop time calculation.
From gettimeofday(2):
The time returned by gettimeofday() is affected by discontinuous jumps in the system time (e.g.,  if  the  system  administrator
       manually changes the system time).  If you need a monotonically increasing clock, see clock_gettime(2).

h2o mainly used on internet servers, so the discontinuous jumps normally not happen for this use case.

but on some system which lack of ntp and internet access and  system time maybe synced with sources link gps,  event loop calculation using gettimeofday may cause jumps issue.

should h2o give a chance that let the user can use **clock_gettime** instead of gettimeofday, eg:
~~~
diff --git a/lib/common/socket/evloop.c.h b/lib/common/socket/evloop.c.h
index 624813a..2d809dd 100644
--- a/lib/common/socket/evloop.c.h
+++ b/lib/common/socket/evloop.c.h
@@ -474,7 +474,14 @@ h2o_evloop_t *create_evloop(size_t sz)
 
 void update_now(h2o_evloop_t *loop)
 {
+#ifdef H2O_EVLOOP_USE_CLOCK_MONOTONIC
+    struct timespec tp;
+    clock_gettime(CLOCK_MONOTONIC, &tp);
+    loop->_tv_at.tv_sec = tp.tv_sec;
+    loop->_tv_at.tv_usec = tp.tv_nsec / 1000;
+#else
     gettimeofday(&loop->_tv_at, NULL);
+#endif
     loop->_now = (uint64_t)loop->_tv_at.tv_sec * 1000 + loop->_tv_at.tv_usec / 1000;
 }
~~~

**Comment by kazuho** [comment]
Thank you for the proposal.

Honestly I am not sure if I like the complexity that is introduced by the the use of `clock_gettime`.

IIUC, the assumption behind the advocates of `clock_gettime` is that programs can have two clocks - one being the wall-clock time (represented by gettimeofday, etc.) and the other being the internal time (represented by clock_gettime). It is assumed that programmers will know which of the two option should be used for every operation related to time, and that requiring them to make such calls is worth the effort. The problem is that I am uncertain if that assumption holds, especially for network applications that would have access to ntp.

This is the reason why we have not used `clock_gettime` until now, and I am not sure if there's new information that warrants us to revisit the decision.

**Comment by chenbd** [comment]
I Added some comments to the original post:

The time jump may affect the timers based on event loop time calculation which cause timed out wrongly, eg:
the server received the http request from client, mark a timeout with 5s and relay the request to other module to handle the request, the module which handle the request should response in 5s, if not, the server reply a timed out to client and mark the request handled.
when the jump happened, the marked timeout with 5s will timed out immediately. 
Normally the server runs well with gettimeofday(2).

IMO the event loop time calculation should not be affected wall time jumps.
Also I looked into libuv and seems the event loop based on monotonic clock.

Please correct me if anything wrong.

**Comment by kazuho** [comment]
> IMO the event loop time calculation should not be affected by wall time jumps.

I think I agree on that. My point is that the deployments can and should be configured so that their wall clocks do not jump. We do not want to see the timestamps found in the log files jump back to the past. We do not want to see cron jobs run twice a day.

That's why slew mode exists, and why it is used in most circumstances (see http://www.ntp.org/ntpfaq/NTP-s-algo.htm#Q-CLOCK-DISCIPLINE), i.e., when the difference is smaller than 128ms.

To summarize, network deployments rely on having access to the correct wall-clock time. And that's especially easy for server-side deployments, which is our primary use-case. That is why I think it is unnecessary _for us_ to use `clock_gettime`.

**Comment by chenbd** [comment]
Take embed devices such as the Automotive Head Unit as example:
The device runs linux, with no network connection, no ntp, with GPS receiver in it which can be used for wall time sync. http server and client run on this device as two process. The device boots with wall time eg: 1970-1-1 00:00:00, when the gps receiver starts working and receive valid time form it, the wall time will be updated. 
Seems it's hard to configure the wall time not jump.

**Comment by kazuho** [comment]
@chenbd Is that the real problem that you are facing? Am I correct to assume that the server starts accepting the connection before the system obtain the correct time?

Anyways, as stated previously, the problem is that maintaining two timestamps is a cost. We do lots of clock-related operations in QUIC, and users can do that in mruby. If they set timers based on the wall-clock time (like, I want to be woken up at 5:00pm, and because it is now 4:59pm, I set the timer to 1 minute), then the clock jumps, things will fail. That means that we'd need to expose API for accessing the `clock_gettime` and require them to correctly use either the wall clock time or the value provided by that API. I do not think that is easy to accomplish.

Therefore, assuming that you'd still want to resolve the issue, I wonder if you could just change `update_now` to use `clock_gettime` instead of `gettimeofday` based on the compiler setting. Then, at least H2O would have one source of time. The downside would be that you'd see the internal value emitted everywhere (incl. logs), but I'd assume that's better than the wall-clock time values in an environment that cannot rely on well-clock time. WDYT?

**Comment by chenbd** [comment]
> @chenbd Is that the real problem that you are facing? Am I correct to assume that the server starts accepting the connection before the system obtain the correct time?
> 
Yes, it happened(with very very low frequency), the client received try again later because the incorrect time out.
 
> Anyways, as stated previously, the problem is that maintaining two timestamps is a cost. We do lots of clock-related operations in QUIC, and users can do that in mruby. If they set timers based on the wall-clock time (like, I want to be woken up at 5:00pm, and because it is now 4:59pm, I set the timer to 1 minute), then the clock jumps, things will fail. That means that we'd need to expose API for accessing the `clock_gettime` and require them to correctly use either the wall clock time or the value provided by that API. I do not think that is easy to accomplish.
> 
IMO with two timestamps  is extra cost, but this depends on use case.
For most of use case of internet related gettimeofday is enough. But is not a good idea to use wall time as timer if wall time can jump. 

> Therefore, assuming that you'd still want to resolve the issue, I wonder if you could just change `update_now` to use `clock_gettime` instead of `gettimeofday` based on the compiler setting. Then, at least H2O would have one source of time. The downside would be that you'd see the internal value emitted everywhere (incl. logs), but I'd assume that's better than the wall-clock time values in an environment that cannot rely on well-clock time. WDYT?

This can fix the issue, but not sure for things like logs, response timing, etc is good idea of not, they need wall time i guess.

**Comment by chenbd** [comment]
> @chenbd Is that the real problem that you are facing? Am I correct to assume that the server starts accepting the connection before the system obtain the correct time?
> 
Yes, it happened(with very very low frequency), the client received try again later because the incorrect time out.
 
> Anyways, as stated previously, the problem is that maintaining two timestamps is a cost. We do lots of clock-related operations in QUIC, and users can do that in mruby. If they set timers based on the wall-clock time (like, I want to be woken up at 5:00pm, and because it is now 4:59pm, I set the timer to 1 minute), then the clock jumps, things will fail. That means that we'd need to expose API for accessing the `clock_gettime` and require them to correctly use either the wall clock time or the value provided by that API. I do not think that is easy to accomplish.
> 
IMO with two timestamps  is extra cost, but this depends on use case.
For most of use case of internet related gettimeofday is enough. But is not a good idea to use wall time as timer if wall time can jump. 

> Therefore, assuming that you'd still want to resolve the issue, I wonder if you could just change `update_now` to use `clock_gettime` instead of `gettimeofday` based on the compiler setting. Then, at least H2O would have one source of time. The downside would be that you'd see the internal value emitted everywhere (incl. logs), but I'd assume that's better than the wall-clock time values in an environment that cannot rely on well-clock time. WDYT?

This can fix the issue, but not sure for things like logs, response timing, etc is good idea of not, they need wall time i guess.
